pkg/discovery/agent.go

```diff
@@ -16,63 +13,84 @@ import (
-	publishBufferSize   int
-	stage               string
-	stopAgent           chan bool
-	stopDiscovery       chan bool
-	stopPublish         chan bool
+	assetCache cache.Cache
+	client     anypoint.Client
+	stopAgent  chan bool
+	discovery  Repeater
+	publisher  Repeater
 }
 
-// New creates a new agent
-func New(cfg *config.AgentConfig, client anypoint.Client) (agent *Agent) {
+// NewAgent creates a new agent
+func NewAgent(cfg *config.AgentConfig, client anypoint.Client) (agent *Agent) {
 	buffer := 5
 	assetCache := cache.New()
-	agent = &Agent{
-		anypointClient:      client,
-		apicClient:          coreagent.GetCentralClient(),
-		apiChan:             make(chan *ServiceDetail, buffer),
+	apiChan := make(chan *ServiceDetail, buffer)
+
+	pub := &publisher{
+		apiChan:     apiChan,
+		stopPublish: make(chan bool),
+		publishAPI:  coreAgent.PublishAPI,
+	}
+
+	svcHandler := &serviceHandler{
 		assetCache:          assetCache,
-		discoveryIgnoreTags: cleanTags(cfg.MulesoftConfig.DiscoveryIgnoreTags),
-		discoveryPageSize:   50,
-		discoveryTags:       cleanTags(cfg.MulesoftConfig.DiscoveryTags),
-		pollInterval:        cfg.MulesoftConfig.PollInterval,
+		freshCache:          cache.New(),
 		stage:               cfg.MulesoftConfig.Environment,
-		stopAgent:           make(chan bool),
-		stopDiscovery:       make(chan bool),
-		stopPublish:         make(chan bool),
+		discoveryTags:       cleanTags(cfg.MulesoftConfig.DiscoveryTags),
+		discoveryIgnoreTags: cleanTags(cfg.MulesoftConfig.DiscoveryIgnoreTags),
+		client:              client,
 	}
 
-	return agent
+	disc := &discovery{
+		apiChan:           apiChan,
+		assetCache:        assetCache,
+		client:            client,
+		discoveryPageSize: 50,
+		pollInterval:      cfg.MulesoftConfig.PollInterval,
+		stopDiscovery:     make(chan bool),
+		serviceHandler:    svcHandler,
+	}
+
+	return newAgent(client, disc, pub, assetCache)
+}
+
+func newAgent(
+	client anypoint.Client,
+	discovery Repeater,
+	publisher Repeater,
+	assetCache cache.Cache,
+) *Agent {
+	return &Agent{
+		client:     client,
+		assetCache: assetCache,
+		discovery:  discovery,
+		publisher:  publisher,
+		stopAgent:  make(chan bool),
+	}
 }
 
 // onConfigChange apply configuration changes
 func (a *Agent) onConfigChange() {
 	cfg := config.GetConfig()
 
 	// Stop Discovery & Publish
-	a.stopDiscovery <- true
-	a.stopPublish <- true
+	a.discovery.Stop()
+	a.publisher.Stop()
 
-	a.stage = cfg.MulesoftConfig.Environment
-	a.discoveryTags = cleanTags(cfg.MulesoftConfig.DiscoveryTags)
-	a.discoveryIgnoreTags = cleanTags(cfg.MulesoftConfig.DiscoveryIgnoreTags)
-	a.apicClient = coreagent.GetCentralClient()
-	a.pollInterval = cfg.MulesoftConfig.PollInterval
-	a.anypointClient.OnConfigChange(cfg.MulesoftConfig)
+	a.client.OnConfigChange(cfg.MulesoftConfig)
+	a.discovery.OnConfigChange(cfg.MulesoftConfig)
 
 	// Restart Discovery & Publish
-	go a.discoveryLoop()
-	go a.publishLoop()
+	go a.discovery.Loop()
+	go a.publisher.Loop()
```

Comment:
The SDK provides a jobs library. I am not sure what happens in these Loop functions but they may be able to implement the Execute() method of the Job interface.  Then you can register the jobs with the SDK.

Using the jobs library gives the added benefit of the SDK being able to stop the jobs when a Status call to any job fails.  Then the jobs library will restart the jobs when all job Ready calls return true.

Maybe not needed now, but something to look at.

Reply:
discovery runs in an interval, so it may be a candidate for the Job interface. Publisher waits on a channel for events published from discovery, so I don't think that one would work in this case.

Reply:
Correct, but you could use the base job type for this now, Vivek and I have talked about the need for a channel job type, but that has not been planned yet.

Reply:
@sagarkal @Atraya21 @chrisccoy What do you guys think of hooking up our 'Job' like functions directly into the sdk?

Reply:
Here is an example of how they do it with the aws agent. https://git.ecd.axway.org/apigov/aws_apigw_discovery_agent/-/blob/master/pkg/aws/pollusageplans.go#L16

Reply:
This is what the sdk provides to register jobs - https://github.com/Axway/agent-sdk/blob/main/pkg/jobs/jobs.go